### PR TITLE
Fix AdClickAttributionLogic leaking NavigationResponse completionBlock

### PR DESCRIPTION
**Pull request by mallexxx**
Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL:  https://app.asana.com/0/1199230911884351/1205322007351871/f
iOS PR: not needed
macOS PR: https://github.com/duckduckgo/macos-browser/pull/1538
What kind of version bump will this require?: Patch

**Description**:
- Fixes async `onProvisionalNavigation` leaking completionBlocks on Navigation Task cancellation
- Added more debug info to `DistributedNavigationDelegate.checkLongDecisionMaking`

**Steps to test this PR**:
1. Set random Integration Tests execution order in Xcode scheme->Tests->Integration Tests Options
2. Run the Integration Tests, validate there‘s no `checkLongDecisionMaking` assertion

### CODE REVIEW

File: Sources/BrowserServicesKit/ContentBlocking/AdClickAttribution/AdClickAttributionLogic.swift [hunk 1]
```diff
@@ -97,6 +97,7 @@ public class AdClickAttributionLogic {
         case .noAttribution:
             self.state = state
         case .preparingAttribution(let vendor, let info, _):
+            cancelAttributedRulesCompilationIfNeeded()
```
Comment (bwaresiak) [inline Sources/BrowserServicesKit/ContentBlocking/AdClickAttribution/AdClickAttributionLogic.swift:100]: `applyInheritedAttribution` should be called only when creating new tab from the context of the attributed tab - it should have no existing state to "cancel" at that point, so if it has, this feels like a possible problem.

File: Sources/BrowserServicesKit/ContentBlocking/AdClickAttribution/AdClickAttributionLogic.swift [hunk 2]
```diff
@@ -196,12 +202,14 @@ public class AdClickAttributionLogic {
                disableAttribution()
            } else if tld.eTLDplus1(host) == vendor {
                os_log(.debug, log: log, "Refreshing navigational duration for attribution")
+               cancelAttributedRulesCompilationIfNeeded()
```
Comment (bwaresiak) [inline Sources/BrowserServicesKit/ContentBlocking/AdClickAttribution/AdClickAttributionLogic.swift:205]: Would that do anything given guard in line 179?

File: Sources/BrowserServicesKit/ContentBlocking/AdClickAttribution/AdClickAttributionLogic.swift [hunk 3]
```diff
@@ -196,12 +202,14 @@ public class AdClickAttributionLogic {
...
                state = .activeAttribution(vendor: vendor,
                                           session: SessionInfo(start: session.attributionStartedAt),
                                           rules: rules)
            }
         } else if tld.eTLDplus1(host) != vendor {
             os_log(.debug, log: log, "Leaving attribution context")
+            cancelAttributedRulesCompilationIfNeeded()
```
Comment (bwaresiak) [inline Sources/BrowserServicesKit/ContentBlocking/AdClickAttribution/AdClickAttributionLogic.swift:212]: Same question as above.

File: Sources/BrowserServicesKit/ContentBlocking/AdClickAttribution/AdClickAttributionLogic.swift [hunk 4]
```diff
@@ -216,8 +224,9 @@ public class AdClickAttributionLogic {
         eventReporting?.fire(.adAttributionActive)
         registerFirstActivity = false
     }
-    
+
     private func disableAttribution() {
+        cancelAttributedRulesCompilationIfNeeded()
```
Comment (bwaresiak) [inline Sources/BrowserServicesKit/ContentBlocking/AdClickAttribution/AdClickAttributionLogic.swift:229]: This is called only if state is active, so I don't think it does anything either.

File: Sources/Navigation/DistributedNavigationDelegate.swift [hunk 5]
```diff
@@ -111,7 +111,8 @@ private extension DistributedNavigationDelegate {
 #endif
 
     /// continues until first non-nil Navigation Responder decision and returned to the `completion` callback
-    func makeAsyncDecision<T>(boundToLifetimeOf webView: WKWebView,
+    func makeAsyncDecision<T>(for actionDebugInfo: some CustomDebugStringConvertible,
+                              boundToLifetimeOf webView: WKWebView,
```
Comment (bwaresiak) [inline Sources/Navigation/DistributedNavigationDelegate.swift:115]: Didn't we agree to disable this kind of warning?

Comment (mallexxx) [reply Sources/Navigation/DistributedNavigationDelegate.swift:115]: As I remember we agreed to make it less intrusive (to stop debugger and not cause assertion) and provide more info. This won't affect CI runners as there's no debugger but may help detecting some issues

**Pull request merged**